packages/nestjs-report/src/services/report.service.ts

```diff
@@ -0,0 +1,134 @@
+import { ReportServiceInterface } from '../interfaces/report-service.interface';
+import { ReportQueryException } from '../exceptions/report-query.exception';
+import { ReportCreateDto } from '../dto/report-create.dto';
+import { ReportStrategyService } from './report-strategy.service';
+import { ReportCreateException } from '../exceptions/report-create.exception';
+import { ReportIdMissingException } from '../exceptions/report-id-missing.exception';
+import { ReportDuplicateEntryException } from '../exceptions/report-duplicated.exception';
+import { ReportGeneratorResultInterface } from '../interfaces/report-generator-result.interface';
+import { mapNonErrorToException } from '@concepta/ts-core';
+
+/**
+ * Service responsible for managing report operations.
+ */
+@Injectable()
+export class ReportService
+  extends BaseService<ReportEntityInterface>
+  implements ReportServiceInterface
+{
+  constructor(
+    @InjectDynamicRepository(REPORT_MODULE_REPORT_ENTITY_KEY)
+    private reportRepo: Repository<ReportEntityInterface>,
+    @Inject(REPORT_STRATEGY_SERVICE_KEY)
+    private reportStrategyService: ReportStrategyService,
+  ) {
+    super(reportRepo);
+  }
+
+  async generate(report: ReportCreateDto): Promise<ReportEntityInterface> {
+    await this.checkExistingReport(report);
+    try {
+      const reportDb = await this.createAndSaveReport(report);
+
+      // trigger report generation
+      this.generateAndProcessReport(reportDb);
+
+      return reportDb;
+    } catch (err) {
+      throw new ReportCreateException(this.metadata.targetName, err);
+    }
+  }
+
+  async fetch(
+    report: Pick<ReportEntityInterface, 'id'>,
+  ): Promise<ReportEntityInterface> {
+    if (!report.id) throw new ReportIdMissingException();
+    const dbReport = await this.reportRepo.findOne({
+      where: {
+        id: report.id,
+      },
+      relations: ['file'],
+    });
+    if (!dbReport) throw new ReportQueryException();
+    return this.addDownloadUrl(dbReport);
+  }
+
+  async done(report: ReportGeneratorResultInterface): Promise<void> {
+    if (!report.id) throw new ReportIdMissingException();
+
+    const updatedReport = await this.reportRepo.findOne({
+      where: { id: report.id },
+    });
+
+    if (!updatedReport) throw new ReportQueryException();
+
+    updatedReport.status = report.status;
+    updatedReport.errorMessage = report.errorMessage || null;
+    updatedReport.file = report.file;
+
+    await this.reportRepo.save(updatedReport);
+  }
+
+  private async generateAndProcessReport(
+    reportDb: ReportEntityInterface,
+  ): Promise<void> {
+    try {
+      const result = await this.reportStrategyService.generate(reportDb);
+      await this.done(result);
+    } catch (err) {
+      const finalError = mapNonErrorToException(err);
+      await this.done({
+        id: reportDb.id,
+        status: ReportStatusEnum.Error,
+        errorMessage: finalError.message,
+      });
+    }
+  }
+
+  private async createAndSaveReport(
```

Comment:
some of these should be `protected` in case somebody extends this service

Reply:
my idea was using private only to better organize the code, the actual method they gonna need to overwrite is the generate/fetch/done methods